### PR TITLE
Ref #2083: jOOR language native support

### LINKED ISSUE #2083: jOOR language native support
https://camel.apache.org/components/latest/languages/joor-language.html

### PR DESCRIPTION
**Pull request by essobedo**
fixes #2083 

## Motivation

The jOOR language can only be used in JVM mode, we would like to support it in native mode too.

## Modifications

* Move the extension to native using the `promote` mojo
* Improve the `support-language` extension to support scripts and expressions parameters
* Add a set of classes allowing to generate an id based on the source code of a script or an expression and the parameters
* Add the class `JoorExpressionConfig` to be able to configure the language at build time 
* Add more integration tests for the extension `csimple` to ensure that predicates and scripts work as expected
* Add a reflection configuration file to the affected integration tests to allow reflective calls via the bean component.

## Result

Expressions and scripts written using the jOOR language are now extracted and compiled at build time.

### CODE REVIEW

File: integration-tests/csimple/src/main/resources/META-INF/native-image/reflect-config.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[
+  {
+    "name" : "java.lang.String",
```
Comment (jamesnetherton) [inline integration-tests/csimple/src/main/resources/META-INF/native-image/reflect-config.json:3]: Nitpick - we could do this with `@RegisterForReflection`.

File: integration-tests/joor/src/main/resources/META-INF/native-image/reflect-config.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[
+  {
+    "name" : "org.apache.camel.quarkus.component.joor.it.JoorBean",
```
Comment (jamesnetherton) [inline integration-tests/joor/src/main/resources/META-INF/native-image/reflect-config.json:3]: Nitpick - we could do this with `@RegisterForReflection`.

File: docs/modules/ROOT/pages/reference/extensions/joor.adoc [hunk 3]
```diff
@@ -39,3 +43,39 @@ Please refer to the above link for usage and configuration details.
...
+| Configuration property | Type | Default
+
+
+|icon:lock[title=Fixed at build time] [[quarkus.camel.joor.single-quotes]]`link:#quarkus.camel.joor.single-quotes[quarkus.camel.joor.single-quotes]`
+
+Indicates whether a jOOR expression can use single quotes instead of double quotes.
+| `boolean`
+| `true`
+
+|icon:lock[title=Fixed at build time] [[quarkus.camel.joor.config-resource]]`link:#quarkus.camel.joor.config-resource[quarkus.camel.joor.config-resource]`
```
Comment (jamesnetherton) [inline docs/modules/ROOT/pages/reference/extensions/joor.adoc:66]: Just an observation - The config descriptions for `config-resource` & `result-type` don't seem to be in the doc.

**Review by jamesnetherton — COMMENTED** [review]
Looks pretty good to me. I left a few (mostly nitpick) comments.

**Comment by jamesnetherton** [comment]
Please create follow up issue(s) for the eventual removal of the temporary workarounds.

File: integration-tests/csimple/src/main/resources/META-INF/native-image/reflect-config.json (hunk 1 above)
Comment (essobedo) [reply integration-tests/csimple/src/main/resources/META-INF/native-image/reflect-config.json:3]: Nice one, thx for sharing 🙏

File: docs/modules/ROOT/pages/reference/extensions/joor.adoc (hunk 3 above)
Comment (essobedo) [reply docs/modules/ROOT/pages/reference/extensions/joor.adoc:66]: Nice catch 👍

**Comment by essobedo** [comment]
> Please create follow up issue(s) for the eventual removal of the temporary workarounds.

Here it is https://github.com/apache/camel-quarkus/issues/4443

File: extensions-support/language/runtime/src/main/java/org/apache/camel/quarkus/support/language/runtime/ExpressionUID.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    protected byte[] asBytes() {
+        StringBuilder source = new StringBuilder();
+        source.append(expression);
+        source.append('.');
+        if (arguments != null) {
+            source.append(Arrays.toString(arguments));
+        }
+        source.append('.');
+        return source.toString().getBytes();
```
Comment (aldettinger) [inline extensions-support/language/runtime/src/main/java/org/apache/camel/quarkus/support/language/runtime/ExpressionUID.java:52]: Not a blocker, still maybe we could avoid to rely on platform's default charset ?

Comment (essobedo) [reply extensions-support/language/runtime/src/main/java/org/apache/camel/quarkus/support/language/runtime/ExpressionUID.java:52]: 👍 I agree

File: integration-tests/joor/src/test/java/org/apache/camel/quarkus/component/joor/it/JoorTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import org.hamcrest.CoreMatchers;
+import org.junit.jupiter.api.Test;
+
+@QuarkusTest
+class JoorTest {
+
+    @Test
+    void loadLanguageJoor() {
+        /* A simple autogenerated test */
+        RestAssured.get("/joor/load/language/joor")
```
Comment (aldettinger) [inline integration-tests/joor/src/test/java/org/apache/camel/quarkus/component/joor/it/JoorTest.java:30]: Is there still value to keep this now ?

**Comment by aldettinger** [comment]
From my side, would need a second reading. Still the frame looks good.

**Review by jamesnetherton — APPROVED** [review]

**Review by aldettinger — APPROVED** [review]
Well done @essobedo , looks a good effort :+1: 

Would you like to open another followup issue to make it even better with more test coverage ?
We generally test any meaningful features documented in camel: https://camel.apache.org/components/3.20.x/languages/joor-language.html

Trying to remind experience from other languages... I think we had seen native issues linked to `bodyAs` like clauses, maybe arrays also...

**Pull request merged**